File: src/jkorn/pvpcore/commands/PvPCommand.php
<?php
/**
 * Created by PhpStorm.
 * User: jkorn2324
 * Date: 2019-04-08
 * Time: 14:03
 */

namespace jkorn\pvpcore\commands;


use jkorn\pvpcore\commands\parameters\Parameter;
use jkorn\pvpcore\commands\parameters\SimpleParameter;
use jkorn\pvpcore\commands\parameters\BaseParameter;
use jkorn\pvpcore\PvPCore;
use pocketmine\command\CommandSender;
use pocketmine\utils\TextFormat;

class PvPCommand extends BaseCommand
{

    /**
     * PvPCommand constructor.
     */
    public function __construct()
    {
        parent::__construct("pvp", "The base command of PvPCore.", "/pvp help");

        $param = new SimpleParameter(Parameter::NO_PERMISSION, "knockback|attackdelay", Parameter::PARAMTYPE_STRING);
        $param = $param->setExactValues(true);

        $this->setParameters([
            [
                new BaseParameter("help", parent::getPermission(), "See all pvp commands.", true)
            ],
            [
                new BaseParameter("set", parent::getPermission(), "Configures the knockback settings of a level.", true),
                new SimpleParameter(Parameter::NO_PERMISSION, "level", Parameter::PARAMTYPE_STRING),
                $param,
                new SimpleParameter(Parameter::NO_PERMISSION, "value", Parameter::PARAMTYPE_ANY)
            ],
            [
                new BaseParameter("enable", parent::getPermission(), "Enables custom knockback in the specified level.", true),
                new SimpleParameter(Parameter::NO_PERMISSION, "level", Parameter::PARAMTYPE_STRING)
            ],
            [
                new BaseParameter("disable", parent::getPermission(), "Disables custom knockback in the specified level"),
                new SimpleParameter(Parameter::NO_PERMISSION, "level", Parameter::PARAMTYPE_STRING)
            ]
        ]);
    }

    /**
     * @param CommandSender $sender
     * @param string $label
     * @param array $args
     * @return bool|mixed
     */
    public function execute(CommandSender $sender, string $label, array $args)
    {
        if (parent::canExecute($sender, $label, $args))
        {
            $param = $args[0];
            switch ($param)
            {
                case "help":
                    $sender->sendMessage($this->getFullUsage());
                    break;
                case "set":
                    $this->executeSetWorld($sender, $args[1], $args[2], $args[3]);
                    break;
                case "enable":
                    $level = $args[1];
                    $this->executeCustomKB($sender, $level, true);
                    break;
                case "disable":
                    $level = $args[1];
                    $this->executeCustomKB($sender, $level, false);
                    break;
            }
        }

        return true;
    }

    /**
     * @param CommandSender $sender
     * @param string $level
     * @param bool $enable
     */
    protected function executeCustomKB(CommandSender $sender, string $level, bool $enable): void
    {

        $worldManager = PvPCore::getWorldHandler();

        if ($worldManager->isWorld($level)) {

            $world = PvPCore::getWorldHandler()->getPvPCWorld($level);

            $format = $enable ? TextFormat::GREEN : TextFormat::RED;

            $enabled = $enable ? "enabled" : "disabled";

            $world = $world->setCustomKBEnabled($enable);

            $worldManager->updateWorld($world);

            $msg = $format . "Custom Knockback has been successfully " . $enabled . "!";

        } else {

            $msg = TextFormat::RED . "Level '$level' does not exist!";
        }

        $sender->sendMessage($msg);
    }

    /**
     * @param CommandSender $sender
     * @param $level
     * @param $setting
     * @param $value
     */
    protected function executeSetWorld(CommandSender $sender, $level, $setting, $value)
    {
        $worldManager = PvPCore::getWorldHandler();

        if ($worldManager->isWorld($level)) {

            $hasUpdated = false;
            $updatedVal = "None";

            switch ($setting) {
                case "kb":
                case "knockback":
                    $updatedVal = "knockback";
                    $value = floatval($value);
                    break;
                case "delay":
                case "attackdelay":
                    $updatedVal = "attack-delay";
                    $value = intval($value);
                    break;
            }

            $world = $worldManager->getPvPCWorld($level);

            if ($updatedVal !== "None") {

                if ($updatedVal === "attack-delay" && PvPCore::canParse($value, true)) {
                    $world = $world->setAttackDelayTime($value);
                    $hasUpdated = true;
                }

                if ($updatedVal === "knockback" && PvPCore::canParse($value, false)) {
                    $world = $world->setKB($value);
                    $hasUpdated = true;
                }
            }

            if ($hasUpdated) {
                $worldManager->updateWorld($world);
                $msg = TextFormat::GREEN . "The level '{$level}' has been successfully updated!";
            } else {
                $msg = TextFormat::RED . "The level '{$level}' failed to update!";
            }

        } else {
            $msg = TextFormat::RED . "Level '{$level}' does not exist!";
        }

        if(isset($msg))
        {
            $sender->sendMessage($msg);
        }

    }
}

File: Nukkit/src/pvpcore/commands/PvPCommand.java
package pvpcore.commands;

import cn.nukkit.command.CommandSender;
import cn.nukkit.utils.TextFormat;
import pvpcore.PvPCore;
import pvpcore.commands.parameters.BaseParameter;
import pvpcore.commands.parameters.Parameter;
import pvpcore.commands.parameters.SimpleParameter;
import pvpcore.misc.PvPCoreUtil;
import pvpcore.worlds.areas.PvPCArea;
import pvpcore.worlds.PvPCWorld;

public class PvPCommand extends BaseCommand {

    public PvPCommand() {
        super("pvp", "The base pvp core command.", "/pvp help");
        Parameter[][] params = new Parameter[][]{
                {new BaseParameter("help", super.getPermission(),"Lists all of the pvp commands.")},
                {new BaseParameter("set", super.getPermission(), "Configures the knockback settings of a level."), new SimpleParameter("level:area", Parameter.PARAMTYPE_STRING).setExact(true), new SimpleParameter("name", Parameter.PARAMTYPE_STRING), new SimpleParameter("knockback:attackdelay", Parameter.PARAMTYPE_STRING).setExact(true), new SimpleParameter("delay:xz-KB", Parameter.PARAMTYPE_ANY, false), new SimpleParameter("y-KB", Parameter.PARAMTYPE_FLOAT, true)},
                {new BaseParameter("enable", super.getPermission(), "Enables custom knockback in a level or area."), new SimpleParameter("level:area", Parameter.PARAMTYPE_STRING).setExact(true), new SimpleParameter("name", Parameter.PARAMTYPE_STRING)},
                {new BaseParameter("disable", super.getPermission(), "Disables custom knockback in a level or area."), new SimpleParameter("level:area", Parameter.PARAMTYPE_STRING).setExact(true), new SimpleParameter("name", Parameter.PARAMTYPE_STRING)}
        };
        setParameters(params);
    }

    @Override
    public boolean execute(CommandSender sender, String label, String[] args) {
        String msg = null;
        if(super.canExecute(sender, label, args)){
            String name = args[0];
            switch(name){
                case "help":
                    msg = getFullUsage();
                    break;
                case "enable":
                    String type = args[1], theName = args[2];
                    executeEnableKB(sender, type, theName, true);
                    break;
                case "disable":
                    type = args[1];
                    theName = args[2];
                    executeEnableKB(sender, type, theName, false);
                    break;
                case "set":
                    int len = args.length;
                    String levelOfArea = args[1], placeName = args[2], valType = args[3];
                    if(len == 6){
                        String xKB = args[4], yKB = args[5];
                        float y = Float.parseFloat(yKB);
                        if(PvPCoreUtil.canParse(xKB, true)){
                            float x = Float.parseFloat(xKB);
                            executeSetKB(sender, levelOfArea, placeName, x, y);
                        } else {
                            msg = getUsageOf(getParamGroupFrom(name), false);
                        }
                    } else {
                        boolean sendMsg = false;
                        if(valType.equalsIgnoreCase("attackdelay")){
                            String attackDelay = args[4];
                            if(PvPCoreUtil.canParse(attackDelay, false)){
                                executeSetDelay(sender, levelOfArea, placeName, Integer.parseInt(attackDelay));
                            } else {
                                sendMsg = true;
                            }
                        } else {
                            sendMsg = true;
                        }

                        if(sendMsg) msg = getUsageOf(getParamGroupFrom(name), false);
                    }
                    break;
                default:
                    msg = getFullUsage();
            }
        }

        if(msg != null) sender.sendMessage(msg);

        return super.execute(sender, label, args);
    }

    private void executeEnableKB(CommandSender sender, String type, String theName, boolean enable) {

        String msg;

        String e = (enable ? TextFormat.GREEN + "enabled" + TextFormat.GRAY : TextFormat.RED + "disabled" + TextFormat.GRAY);
        if (type.equalsIgnoreCase("level")) {
            if(PvPCore.worldHandler.doesWorldExist(theName)){
                PvPCWorld world = PvPCore.worldHandler.get(theName);
                world = world.setKBEnabled(enable);
                PvPCore.worldHandler.update(world);
                msg = TextFormat.GRAY + "Custom KB for the level '" + TextFormat.BLUE + world.getWorldName() + TextFormat.GRAY + "' has been successfully " + e + "!";
            } else {
                msg = TextFormat.RED + "The level '" + theName + "' does not exist!";
            }
        } else {
            if(PvPCore.areaHandler.doesAreaExist(theName)){
                PvPCArea area = PvPCore.areaHandler.getAreaByName(theName);
                area = area.setAreaKBEnabled(enable);
                PvPCore.areaHandler.update(area);
                msg = TextFormat.GRAY + "Custom KB for the area '" + TextFormat.BLUE + area.getName() + TextFormat.GRAY + "' has been successfully " + e + "!";
            } else {
                msg = TextFormat.RED + "The arena '" + theName + "' does not exist!";
            }
        }

        sender.sendMessage(msg);

    }

    protected void executeSetDelay(CommandSender sender, String levelOrArea, String placeName, int attackDelay) {

        String msg;

        if(levelOrArea.equalsIgnoreCase("level")){
            if(PvPCore.worldHandler.doesWorldExist(placeName)){
                if(attackDelay <= 0){
                    msg = TextFormat.RED + "The value has to be greater than 0!";
                } else {
                    PvPCWorld world = PvPCore.worldHandler.get(placeName);
                    world = world.setAttackDelay(attackDelay);
                    PvPCore.worldHandler.update(world);
                    msg = TextFormat.GRAY + world.getWorldName() + " has been successfully" + TextFormat.GREEN + " updated" + TextFormat.GRAY + "!";
                }
            } else {
                msg = TextFormat.RED + "The level '" + placeName + "' does not exist!";
            }
        } else {
            if(PvPCore.areaHandler.doesAreaExist(placeName)){
                if(attackDelay <= 0){
                    msg = TextFormat.RED + "The value has to be greater than 0!";
                } else {
                    PvPCArea area = PvPCore.areaHandler.getAreaByName(placeName);
                    area = area.setAttackDelay(attackDelay);
                    PvPCore.areaHandler.update(area);
                    msg = TextFormat.GRAY + area.getName() + " has been successfully" + TextFormat.GREEN + " updated" + TextFormat.GRAY + "!";
                }
            } else {
                msg = TextFormat.RED + "The arena '" + placeName + "' does not exist!";
            }
        }

        sender.sendMessage(msg);
    }

    protected void executeSetKB(CommandSender sender, String levelOfArea, String name, float xKb, float yKb){
        String msg = null;
        if(levelOfArea.equalsIgnoreCase("level")){
            if(PvPCore.worldHandler.doesWorldExist(name)){
                if(xKb <= 0 || yKb <= 0){
                    msg = TextFormat.RED + "The value has to be greater than 0!";
                } else {
                    PvPCWorld world = PvPCore.worldHandler.get(name);
                    world = world.setKB(xKb, yKb);
                    PvPCore.worldHandler.update(world);
                    msg = TextFormat.GRAY + world.getWorldName() + " has been successfully" + TextFormat.GREEN + " updated" + TextFormat.GRAY + "!";
                }
            } else {
                msg = TextFormat.RED + "The level '" + name + "' does not exist!";
            }
        } else {
            if(PvPCore.areaHandler.doesAreaExist(name)){
                if(xKb <= 0 || yKb <= 0){

                } else {
                    PvPCArea area = PvPCore.areaHandler.getAreaByName(name);
                    area = area.setKB(xKb, yKb);
                    PvPCore.areaHandler.update(area);
                    msg = TextFormat.GRAY + area.getName() + " has been successfully" + TextFormat.GREEN + " updated" + TextFormat.GRAY + "!";
                }
            } else {
                msg = TextFormat.RED + "The arena '" + name + "' does not exist!";
            }
        }
        if(msg != null) sender.sendMessage(msg);
    }

}
